### DISCUSSION TITLE
Fitview of Nodes when the size of canvas changes.

### DISCUSSION THREAD

**Discussion by Santoshsohani** [xyflow/xyflow#4635] [Q&A]
I have a problem where the initial size of the main canvas is supposed to be `100vh` after click of a button the size of canvas becomes `50vh`. Now I need to fitview the nodes when the canvas shrinks. Is there any approach.

**Comment by veksen** [comment]
I did something similar with a ResizeObserver on an element that wraps the canvas (or itself):
(make sure you have a `ReactFlowProvider` wrapping the component you use `useReactFlow` in)

```
const { fitView } = useReactFlow();

React.useEffect(() => {
  const tree = treeRef.current;

  const observer = new ResizeObserver(() => {
    requestAnimationFrame(() => {
      fitView();
    });
  });

  if (tree) {
    observer.observe(tree);
  }

  return () => {
    if (tree) {
      observer.unobserve(tree);
    }
  };
}, [fitView]);
```

**Comment by ckpearson** [comment]
Hi, you could also apply the `fitView` attribute to the flow element to cause it to automatically fit the view on render, you may need to trigger a re-render if the layout changes but that's likely happening anyway if you're driving your height change via state

**Reply by bcakmakoglu** [reply]
That would basically mean forcing a full re-render of ReactFlow though which seems inefficient imo, or did I misunderstand your proposal here?

**Reply by ckpearson** [reply]
it would, but it's going to re-render when nodes change anyway, and if you've got any `useState` hooks higher in your component tree the flow is going to re-render anyway, so it's probably fairly negligible.

**Reply by bcakmakoglu** [reply]
I'm not quite seeing how that would be a *full remount* which the `fitView` prop would need to be triggered, so unless you're using a `key` or something I don't see this working 😅

**Reply by ckpearson** [reply]
We do indeed use a key for a similar thing, basically we needed Dagre layout on-render + node addition as well as fitView, so we derive a memoized key value from changes to our nodes collection and provide that to the flow editor, performance has been fine tbh.

**Reply by bcakmakoglu** [reply]
Sure it works, although I'd just prefer to listen to the viewport size change, do a re-layout and fit view instead of forcing a remount on ReactFlow and all my custom nodes / edges, which might have unintended effects as well, wouldn't it? 
What if we have some `useEffect(() => { /* do something on mount */ }, [])` that shouldn't re-run because of a size change? 
Sure, we can work around that with a state (or not, now that I think about it *if* a remount happens the state wouldn't persist either, right?) but it just seems like forcing unexpected behavior and/or annoying bugs.

(although I'm no expert in React so tbh not sure if a key change would force a full remount on everything, so if that's wrong correct me on that please 😅)

**Reply by ckpearson** [reply]
Yeah passing a fresh key to the flow component would trigger a re-mount of the flow, but tbh so long as the data you're supplying is  consistent it shouldn't be any more expensive than the initial render, if you're concerned about this happening for every size change event then you would probably want to debounce the size changes and have the re-mount and fitView to occur after the resize has finished, a common pattern to prevent the layout causing a "flash" as the render occurs is to have the opacity of the flow container switch to 0 and then 1 to hide that happening, it's usually so fast you don't really notice anything it just looks like something has loaded in.

**Reply by bcakmakoglu** [reply]
I'm not disputing that *it would work* and expenses aren't my main concern either after thinking about it with the re-mounting in mind.
Using a key would potentially cause unintended side-effects some people wouldn't be too familiar with and
I don't see a particular upside to using a key rather than triggering a layout and fit view in an effect 😅

There was a discussion about this going on here: https://github.com/xyflow/xyflow/discussions/4575#discussioncomment-10410413

Basically it boiled down to
```ts
  useEffect(() => {
    setTimeout(() => {
      fitView();
    }, 5);
  }, [sidebarVisible, fitView]);
```

Although currently `fitView` needs a timeout to be able to fit properly after a size change (a sidebar toggling in this example), which is an issue I've reported already, so if that were fixed a timeout shouldn't be necessary either (which would be the one thing I don't like about this particular implementation).

**Reply by ckpearson** [reply]
Yeah we also had the fitView in a timeout / `requestAnimationFrame` thing but then of course you end up with the "jump", so also ended up doing the trick where you set the opacity to 0 and then bring it in to 1 when the `onInit` event fires from the flow.

**Reply by bcakmakoglu** [reply]
Once https://github.com/xyflow/xyflow/issues/4652 is fixed a timeout shouldn't be necessary anymore :)
